crates/symbolicator/src/services/cacher.rs

```diff
@@ -356,4 +352,38 @@ impl<T: CacheItemRequest> Cacher<T> {
             })
             .await
     }
+
+    /// Computes an item by loading from or populating the cache.
+    ///
+    /// The actual computation is deduplicated between concurrent requests. Finally, the result is
+    /// inserted into the cache and all subsequent calls fetch from the cache.
+    ///
+    /// The computation itself is done by [`T::compute`](CacheItemRequest::compute), but only if it
+    /// was not already in the cache.
+    ///
+    /// # Errors
+    ///
+    /// Cache computation can fail, in which case [`T::compute`](CacheItemRequest::compute)
+    /// will return an error of type [`T::Error`](CacheItemRequest::Error).  When this
+    /// occurs the error result is returned, **however** in this case nothing is written
+    /// into the cache and the next call to the same cache item will attempt to re-compute
+    /// the cache.
+    pub async fn compute_memoized(&self, request: T) -> Result<Arc<T::Item>, Arc<T::Error>> {
+        let name = self.config.name();
+        let key = request.get_cache_key();
+
+        // cache_path is None when caching is disabled.
+        let cache_path = self.get_cache_path(&key);
+        if let Some(ref path) = cache_path {
+            if let Some(item) = self.lookup_cache(&request, &key, path)? {
+                return Ok(Arc::new(item));
+            }
+        }
+
+        // A file was not found. If this spikes, it's possible that the filesystem cache
+        // just got pruned.
+        metric!(counter(&format!("caches.{}.file.miss", name)) += 1);
+
+        Ok(self.spawn_computation(request.clone(), cache_path).await?)
```

Comment:
iiuc you can run into a race condition here where you checked the cache key, didn't find anything but a computation is already running, then spawn the computation but the previous computation already completed and now you're computing it twice.  I think this is something the old version avoided?
(sorry, didn't realise this yesterday when we were talking about this)

Reply:
I think that is in theory possible, but wouldn’t worry about it too much, since these things are idempotent.
But we can do an additional cache fetch attempt inside of `compute` just in case, wdyt?

Reply:
yes, that's probably a reasonable solution